File: homework3/intersections.py
import matplotlib.pyplot as plt
import numpy as np
from newton import solve

x = np.linspace(-10, 10, 1000)
y1 = x * np.cos(np.pi * x)
y2 = 1 - 0.6 * (x * x)

plt.figure(1)
plt.clf()
plt.plot(x, y1, 'b-')
plt.plot(x, y2, 'r-')

def fun(x):
    f = x * np.cos(np.pi * x) - 1 + 0.6 * x**2
    fp = np.cos(np.pi * x) - x * np.sin(np.pi * x) * np.pi + 1.2 * x
    return f, fp

guesses = [-2.2, -1.6, -0.8, 1.4]
for x0 in guesses:
    print " "  # blank line
    x, iters = solve(fun, x0)
    print "solve returns x = %22.15e after %i iterations " % (x,iters)
    fx1 = x * np.cos(np.pi * x)
    fx2 = 1 - 0.6 * x**2
    print "the value of f(x) is %22.15e" % (fx1 - fx2)
    assert abs(fx1 - fx2) < 1e-14, "*** Unexpected result: x = %22.15e"  % x

    plt.plot(x, fx1, 'ko')

plt.savefig('intersections.png')

File: homework3/intersections.f90
program intersections

    use newton, only: solve

    implicit none
    integer :: iters, itest
    real(kind=8) :: x, x0, fx
    real(kind=8) :: x0vals(4)
    real(kind=8), external :: fun, funp
    logical :: debug

    x0vals = (/-2.2, -1.6, -0.8, 1.4/)
    debug = .true.

    do itest=1,4
        x0 = x0vals(itest)
        call solve(fun, funp, x0, x, iters, debug)

        print 11, x, iters
11      format('sovler returns x = ', es22.15 , ' after', i3, 'iterations')

        fx = fun(x)
        print 12, fx
12      format('the value of f(x) is ', es22.15)

        if (abs(fx) > 1d-14) then
            print 13, x
13          format('*** Unexpected results: x = ', es22.15)
        endif
    enddo

end program intersections

real(kind=8) function fun(x)
    implicit none
    real(kind=8), intent(in) :: x
    real(kind=8) :: pi = 3.14159265358979
    fun = x * cos(pi * x) - 1 + 0.6 * x**2
end function

real(kind=8) function funp(x)
    implicit none
    real(kind=8), intent(in) :: x
    real(kind=8) :: pi = 3.14159265358979
    funp = cos(pi * x) - x * sin(pi * x) * pi + 1.2 * x
end function
